snuba/query/formatters/tracing.py

```diff
@@ -1,169 +1,158 @@
-        return ret
+    selects = ",\n".join(
+        [
+            f"{e.expression.accept(eformatter)} |> {e.name}"
+            for e in query.get_selected_columns()
+        ]
+    )
+    select_str = f"SELECT\n{selects}" if selects else ""
+
+    from_strs = [
+        "FROM",
+        *_indent_str_list(
+            TracingQueryFormatter().visit(query.get_from_clause()), levels=1
+        ),
+    ]
+
+    groupbys = ",\n".join([e.accept(eformatter) for e in query.get_groupby()])
+    groupby_str = f"GROUPBY\n{groupbys}" if groupbys else ""
+
+    orderbys = ",\n".join(
+        [
+            f"{e.expression.accept(eformatter)} {e.direction.value}"
+            for e in query.get_orderby()
+        ]
+    )
+    orderby_str = f"ORDER_BY\n{orderbys}" if orderbys else ""
 
-    def visit_curried_function_call(
-        self, exp: CurriedFunctionCall
-    ) -> Sequence[TExpression]:
-        ret: List[TExpression] = [exp.alias] if exp.alias else []
+    str_list = [select_str, *from_strs, groupby_str, orderby_str]
 
-        ret.extend(
-            [exp.internal_function.accept(self), self.__visit_params(exp.parameters)]
+    array_join = query.get_arrayjoin()
+    if array_join:
+        str_list.append(f"ARRAYJOIN\n{array_join.accept(eformatter)}")
+    condition = query.get_condition()
+    if condition:
+        str_list.append(f"WHERE\n{condition.accept(eformatter)}")
+    having = query.get_having()
+    if having:
+        str_list.append(f"HAVING\n{having.accept(eformatter)}")
+    limitby = query.get_limitby()
+    if limitby:
+        str_list.append(
+            f"LIMIT {limitby.limit} BY {limitby.expression.accept(eformatter)}"
         )
-        return ret
-
-    def visit_argument(self, exp: Argument) -> str:
-        return self.__aliasify_str(exp.name, exp.alias)
-
-    def visit_lambda(self, exp: Lambda) -> Sequence[TExpression]:
-        ret: List[TExpression] = [exp.alias] if exp.alias else []
-        ret.extend([list(exp.parameters), exp.transformation.accept(self)])
-        return ret
+    limit = query.get_limit()
+    if limit:
+        str_list.append(f"  LIMIT {limit}")
+    offset = query.get_offset()
+    if offset:
+        str_list.append(f"  OFFSET {offset}")
+    # The StringifyVisitor formats expression with newlines but we
+    # need them as a list of strings (to avoid truncation)
+    # therefore as a last step we join our list of strings into one string,
+    # and then split it again by newlines to go from this:
+
+    # ['SELECT\n  c1,  ev.c AS `_snuba_c1`\n  f1,  f(\n    ev.c2\n  ) AS `_snuba_f1`',
+    #  'FROM',
+    #  "  ENTITY(events) AS `ev`",
+    # ]
+
+    # To this:
+
+    # ['SELECT',
+    #  '  c1,  ev.c AS `_snuba_c1`',
+    #  '  f1,  f(',
+    #  '    ev.c2',
+    #  '  ) AS `_snuba_f1`',
+    #  '',
+    #  '',
+    #  'FROM',
+    #  "  ENTITY(events) AS `ev`",
+    return "\n".join(str_list).rstrip().split("\n")
```

Comment:
What was the reason for splitting the string produced by the `StringifyVisitor` instead of making the visitor returning a list of strings? Splitting based on the newline character introduces coupling between the implementation of this formatter and the implementation of the `StrigifyVisitor` while if we made the visitor return a list of strings we would depend only on the interface.

Reply:
The reason I made it a string is because the StringifyVisitor code is much more readable if you're composite a string  rather than a list of strings where every line is a string. I do agree that the interface is cleaner if it is a list of strings. I can rewrite it as a list of strings and we can see which reality we prefer